### PR TITLE
feat: added typescript support and updated docs accordingly

### LINKED ISSUE #47: Property '$strapi' does not exist on type
Hi, I have a problem with  `$strapi` in `pages/index.vue`. I already add  `@nuxtjs/strapi` in `nuxt.config.js` modules but why I got this error...

Capture

Or

Capture2

Can you guys give me some advice? thank.

### LINKED ISSUE #61: Typescript support
# What is Expected: No Error

# What is happening
I wanted to use this module with nuxt, typescript and vue composition api but there are no type declarations for it, so I started writing my own, But still there is an error telling me that there is no $strapi on context when I want to use $strapi in a middleware.
The exact error: 
{
	"resource": "/project/middleware/auth.ts",
	"owner": "typescript",
	"code": "2339",
	"severity": 8,
	"message": "Property '$strapi' does not exist on type 'Context'.",
	"source": "ts",
	"startLineNumber": 4,
	"startColumn": 29,
	"endLineNumber": 4,
	"endColumn": 36
}

I want to be able to use this module in a nuxt-typescript project without errors.

# What I think would fix the error
I think rewriting the module in typescript or just providing @types folder would fix the issue.
I myself would be interested in helping.

# My code
```js
import { Middleware } from '@nuxt/types'

const auth: Middleware = ({ $strapi, redirect, route }) => {
  if (!$strapi.user) {
    return redirect(`/login?next=${route.path}`)
  }
}

export default auth
```

### PR DESCRIPTION
**Pull request by LuckeeDev**
## Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description
Why? I implemented this changes in a personal project and, after finding them extremely useful, I decided to contribute to the repo to make TypeScript support available out-of-the-box for the library.

Added TypeScript support:
- Now the library has an `index.d.ts` which extends Vue and Context interfaces with [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) to provide typings for the module.

Updated the docs:
- Added a "Usage with TypeScript" page to describe the setup and usage with this new feature;
- Moved "Using a proxy" to the last section of the docs and inserted the TypeScript page at the 7th place;
- Added some lines about TypeScript in the `index.md`.

Dependencies:
- Added `@nuxt/types` in the `devDependencies` to provide the ability to extend the `Context` interface.

Resolves #61. Resolves #47.

## Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes (if not applicable, please state why)
> Not applicable because changes only regard type definitions.

### CODE REVIEW

**Comment by Atinux** [comment]
Awesome work @LuckeeDev 🙌 

One question, should we also add the `types` field in the `package.json`?
https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#including-declarations-in-your-npm-package

**Comment by LuckeeDev** [comment]
> Awesome work @LuckeeDev 🙌
> 
> One question, should we also add the `types` field in the `package.json`?
> https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#including-declarations-in-your-npm-package

Quoting from the link you provided:

> Also note that if your main declaration file is named index.d.ts and lives at the root of the package (next to index.js) you do not need to mark the "types" property, though it is advisable to do so.

So yes, I'm going to add it right now and then commit.

File: lib/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+  $cookies: NuxtCookies;
+}
+
+declare module 'vue/types/vue' {
+  interface Vue {
+    $strapi: NuxtStrapi;
+  }
+}
+
```
Comment (danielroe) [inline lib/index.d.ts:155]:
```suggestion

declare module 'vuex/types/index' {
  interface Store<S> {
    $strapi: NuxtStrapi;
  }
}

```

**Review by danielroe — COMMENTED** [review]
Looking good to me :heart:

File: docs/content/en/typescript.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+}
+```
+
+## Usage
+
+You now have access to `this.$strapi` inside your components and to `ctx.$strapi` inside 
+`asyncData`, `fetch`, `plugins`, `middlewares` and `nuxtServerInit`.
+
+### In component methods
+```vue
```
[the remark this answers is not in the document]
Comment (LuckeeDev) [reply docs/content/en/typescript.md:30]: Great suggestion!

**Comment by LuckeeDev** [comment]
Thanks @danielroe!
I just implemented and committed all of your three suggestions.

File: lib/index.d.ts (hunk 1 above)
Comment (LuckeeDev) [reply lib/index.d.ts:155]: Should I add an example in the docs implementing the new Store typings?

Comment (LuckeeDev) [reply lib/index.d.ts:155]: After re-reading this, I don't think I understand how this could be used: would you mind to explain?

Comment (danielroe) [reply lib/index.d.ts:155]: Within a store action, you can access `this.app.$strapi` - so this declaration lets us know about that.

File: docs/content/en/index.md [hunk 3]
```diff
@@ -8,13 +8,17 @@ features:
...
 ---
 
 <img src="/preview.png" class="light-img" width="1280" height="640" alt=""/>
 <img src="/preview-dark.png" class="dark-img" width="1280" height="640" alt=""/>
 
 This module is a wrapper around [Strapi](https://strapi.io/) for [NuxtJS](https://nuxtjs.org).
 
+If you want to know how to use this module with TypeScript,
+check out [this page](https://strapi.nuxtjs.org/typescript) of the docs.
+
```
Comment (danielroe) [inline docs/content/en/index.md:21]: I think this is probably not necessary in this file given that we have a link on the sidebar. You could add a comment to `setup.md` at the end though.

**Review by danielroe — COMMENTED** [review]
Sorry - missed out a comment previously. Apart from this, I think it looks good to go :-D

File: docs/content/en/index.md (hunk 3 above)
Comment (LuckeeDev) [reply docs/content/en/index.md:21]: Ok, I removed it

File: lib/index.d.ts (hunk 1 above)
Comment (LuckeeDev) [reply lib/index.d.ts:155]: Oh ok thanks. Should we also add a `NuxtAppOptions` declaration like below?
```vue
declare module '@nuxt/types' {
  interface NuxtAppOptions {
    $strapi: NuxtStrapi;
  }
}
```

Comment (danielroe) [reply lib/index.d.ts:155]: Good catch :+1:

Comment (LuckeeDev) [reply lib/index.d.ts:155]: Added and committed 👍

**Comment by LuckeeDev** [comment]
I think it's ready now!

File: lib/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
+import Vue from 'vue';
+import { Context, NuxtAppOptions } from '@nuxt/types';
```
Comment (danielroe) [inline lib/index.d.ts:2]: You shouldn't need this import (or the dependency on `@nuxt/types`)

Comment (LuckeeDev) [reply lib/index.d.ts]: Right, removed.

**Review by danielroe — APPROVED** [review]

**Pull request merged**